src/globus_sdk/services/auth/consents.py

```diff
@@ -0,0 +1,311 @@
+"""
+This module provides convenience data structures to model and interact with
+    Globus Auth consents.
+
+The resources defined herein are:
+    * ``Consent``
+        * A data object modeling a user's grant for a client to perform some scoped
+            operation on their behalf.
+        * This grant is conditional on the invocation path leading to the client's
+            attempted operation being initiated through a chain of similarly scoped
+            operations (consents) defined in the "dependency_path".
+    * ``ConsentTree``
+        * A tree composed of Consent nodes with edges modeling the dependency
+            relationships between them.
+        * A `meets_scope_requirements` method is defined to check whether a scope
+            requirement, including dependent scope requirements, is satisfied by the
+            tree.
+    * ``ConsentForest``
+        * A collection of all ConsentTrees for a user rooted under a particular client
+            (the client that initiated the request for consents).
+        * A `meets_scope_requirements` method is defined to check whether a scope
+            requirement, including dependent scope requirements, is satisfied by any
+            tree in the forest.
+"""
+
+from __future__ import annotations
+
+import json
+import typing as t
+from dataclasses import dataclass
+from datetime import datetime
+
+from globus_sdk import Scope
+from globus_sdk._types import UUIDLike
+
+from .errors import ConsentParseError
+
+
+@dataclass
+class Consent:
+    """
+    Consent Data Object
+
+    This object models:
+        * A grant which a user has provided for a client to perform a particular
+            scoped operation on their behalf.
+        * The consent is conditional on the invocation path leading to the client's
+            attempted operation being initiated through a chain of similarly scoped
+            operations (consents) defined in the "dependency_path".
+    """
+
+    client: UUIDLike
+    scope: UUIDLike
+    scope_name: str
+    id: int
+    effective_identity: UUIDLike
+    # A list representing the path of consent dependencies lading from a "root consent"
+    #   to this. The last element of this list will always be this consent's ID.
+    # Downstream dependency relationships may exist but will not be defined here.
+    dependency_path: list[int]
+    created: datetime
+    updated: datetime
+    last_used: datetime
```

Comment:
> I'd prefer to draw a firmer line that we don't do more work to the API response than is strictly necessary for our purposes

At a core level I don't necessarily agree but I'm fine with this being the standard in SDK APIs for now.

In this instance though we're not modifying an API response. We're providing a separate class hierarchy where users can ask us explicitly to pass their unmodified API response to a dataclass heirarchy that loads it into python maintained helper data structures.

Reply:
I think this is part of a broader matter of "failing when we didn't need to fail".
For example, if these fields are or become nullable, `datetime` parsing will _fail_, whereas marking these attributes as `str` will merely make the type information incorrect (which is _much_ lower impact -- not a non-issue, but typing is only used by a small class of our users, mostly within Globus, anyway).

I also dislike the fact that this is an informationally lossy conversion, although the information carried is not typically semantically significant:
```
2022-01-01T00:00:00+00:00
2022-01-01T00:00:00-00:00
2022-01-01T00:00:00Z
```
all parse to the same object.

---

My read is that this is an application of the Robustness Principle. Doing the parse has a potential to do harm and I don't see it providing us with any benefits.

I get what you're saying about it being part of the externalized model we provide to users -- that these data are timestamps -- but I want us to be very conservative about doing any additional work in the SDK, even in an opt-in context like this one, since it introduces additional failure points. IMO the better model is to expose users to the truth of the underlying API, which is that they'll get these data as strings, and it's relatively easy to parse them if they want.

Reply:
>  ```
>   2022-01-01T00:00:00+00:00
>   2022-01-01T00:00:00-00:00
>   2022-01-01T00:00:00Z
>   ```
>   all parse to the same object.

They should?

If someone's concerned about differentiating between different serializations of the same conceptual object, they shouldn't be dealing with a "loaded into python data structure" version of their data. They should be dealing with the json response which they have direct access to.

The whole idea behind this class hierarchy is to provide convenient "here's a helpful representation of this stuff in python" view of the data extracted from the "pure serialized json response we got from the service". By using this load my data path, they've specifically requested the hand-holdy path.

---

> My read is that this is an application of the Robustness Principle. Doing the parse has a potential to do harm and I don't see it providing us with any benefits.

It has the benefit of exposing datetimes as python datetimes. If you'd like to work with the last updated time returned by this API, you don't need to look up documentation describing seralization format, you get handed a thing in the standard datetime format in python.

---

It's true that a future change could make this field nullable and then we'd need to update our class to reflect that. If that happens then we'd simply make a change to this class in the SDK to make this field optional and push out a minor version release.

Reply:
We talked about this fairly extensively and agreed that most of the dangers and worries we have stem from clarity around what our contract is with the upstream dev team. By running the proposed implementation past the Auth team, we established that this is what we're planning, and that it's known to the impacted service maintainers and not flagged as an obvious problem.

We're also considering a future implementation which evaluates attributes lazily, rather than eagerly, but getting dataclass-like simplicity, mypy-friendly behavior, and runtime checks and conversions is all a bit... much. We'll need to think through how to approach this more.